src/client/Client.cc

```diff
@@ -6865,8 +6865,11 @@ void Client::tick()
       check_caps(in, CHECK_CAPS_NODELAY);
   }
 
-  if (!mount_aborted)
+  /* Do not send the metrics until the MDS rank is ready */
+  if (!mount_aborted && mdsmap->is_active((mds_rank_t)0)) {
     collect_and_send_metrics();
+  }
```

Comment:
This way it would be hard to debug why the metrics data wasn't sent. Maybe @joscollin meant to add this patch as well as the existing two checks i.e.:
```
  if (!have_open_session((mds_rank_t)0)) {
    ldout(cct, 5) << __func__ << ": no session with rank=0 -- not sending metric"
                  << dendl;
    return;
  }
  auto session = _get_or_open_mds_session((mds_rank_t)0);
  if (!session->mds_features.test(CEPHFS_FEATURE_METRIC_COLLECT)) {
    ldout(cct, 5) << __func__ << ": rank=0 does not support metrics" << dendl;
    return;
  }
```
inside `if(!mount_aborted) {}` ? Cause if none of these satisfies, then then there is no need to even call `Client::collect_and_send_metrics()` right?

Reply:
I will add one helper to do the whole checks.

Reply:
Wait. It seems the `Client::collect_and_send_metrics()` will collect and send non-global metrics ? If that's true then it maybe not using the `rank 0` for that. And then only the `collect_and_send_global_metrics()` will use the `rank 0`.

```
 6969 void Client::collect_and_send_metrics() {
 6970   ldout(cct, 20) << __func__ << dendl;
 6971  
 6972   ceph_assert(ceph_mutex_is_locked_by_me(client_lock));
 6973  
 6974   // right now, we only track and send global metrics. its sufficient
 6975   // to send these metrics to MDS rank0.
 6976   collect_and_send_global_metrics();  
 6977 }

```

If that's is true we should do all the checked in the `collect_and_send_global_metrics()` instead as https://github.com/ceph/ceph/commit/a2be5ec91462a9ed665da803c0f6c2f44fa551fb.

Reply:
I just revert it back to https://github.com/ceph/ceph/commit/a2be5ec91462a9ed665da803c0f6c2f44fa551fb.

Reply:
This makes sense. So, this also means collecting and sending global metrics must be done only by rank 0?

Reply:
The `rank 0` is responsible for collecting the metrics. And if we send the metrics to other ranks then those ranks will forward the metrics to `rank 0` anyway. So in client side we just send to `rank 0`.

Reply:
got it, thanks xiubo

Reply:
okay. I think still the change could move down after `have_open_session`, which is better. It matters only if we have an open session. wdyt?

Reply:
the session(s) will be open in `up:reconnect`, and in this state clients can resend the requests that were replied but not yet journaled. Also if you look at `have_open_session`, it doesn't only look for `open` sessions but `state` sessions too. So I think the placement of the code is better at the top.